e2e/nomostest/registry_server.go

```diff
@@ -97,32 +98,136 @@ func setupRegistry(nt *NT) error {
 	return nil
 }
 
-// setupRegistryClient handles registry authentication.
-// This is necessary on each test because gcloud auth tokens expire after 1h and
-// some test suites take longer than 1h.
+// setupRegistryClient handles registry authentication and port-forwarding.
+//
+// Re-auth is necessary on each test because gcloud auth tokens expire after 1h
+// and some test suites take longer than 1h.
+//
+// Port-forwarding is configured for local registries so that if the registry
+// pod crashes or is rescheduled the port-forwarding will be restarted after it
+// becomes healthy again. Then the client will re-login and re-push all the
+// images and charts that were previously pushed during the same test.
+//
+// For non-local registries, only login, logoiut, and cleanup are performed.
+//
+// Helm client is only configured if RequireHelmProvider is true.
+// OCI client is only configured if RequireOCIProvider is true.
 func setupRegistryClient(nt *NT, opts *ntopts.New) error {
-	nt.T.Cleanup(func() {
-		if opts.RequireHelmProvider || opts.RequireLocalHelmProvider {
-			if err := nt.HelmProvider.Logout(); err != nil {
-				nt.T.Error(err)
+
+	// For local registries, set up port-forwarding to the in-cluster registry server.
+	if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local || opts.RequireHelmProvider && *e2e.HelmProvider == e2e.Local {
+		// The local registry uses ephemeral storage. So login and re-push all
+		// the oci images and helm charts after the registry recovers from a crash.
+		// Before the test, nt.ociImages & nt.helmPackages will be empty.
+		var onReadyCallbacks []func(int, string)
+		if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local {
+			provider := nt.OCIProvider.(*registryproviders.LocalOCIProvider)
+			setupFn := func(newPort int, podName string) {
+				if err := provider.Login(); err != nil {
```

Comment:
Does it need to login with local provider?

Reply:
Not today, but it shouldn't need to know that the local Login is a no-op.

Reply:
FWIW the in-cluster registry deployment already supports basic auth for testing, but surfaces a separate unauthenticated port for the test framework to push to for convenience. So this may be an implementation detail, but I don't think the test scaffolding should ever need to worry about it

Reply:
If we use auth, can we skip the port-forwarding?

Reply:
No, these are separate problems. The port forwarding is so that the test scaffolding can push to the in-cluster registry

Reply:
Similar to how the test scaffolding pushes to the in-cluster git server using the local port forward

Reply:
Oh right, because we didn't want to set up ingress and we want to delay calls when it's unhealthy.

Reply:
This Login is now moved into `ProxiedRegistryProvider.Restore`. So `portForwardRegistryServer` doesn't need to know whether login is necessary or not or how to do it.